### PR TITLE
Implement rate limiting

### PR DESCRIPTION
**Pull request by kalabukdima**
Replace counting the number of requests per epoch with limiting RPS by using the token bucket algorithm, as described [here](https://github.com/subsquid/specs/blob/main/network-rfc/06_compute_units_allocation.md#rate-limiting).

On each query, if the limit has already been hit or the last token has been used, the `retry_after_ms` hint is added to the response. It tells the client that no more queries can be processed until that duration elapses.

### CODE REVIEW

File: src/gateway_allocations/rate_limiter.rs [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+pub struct RateLimiter {
+    operators: HashMap<Address, Bucket>,
+    operator_by_gateway_id: HashMap<PeerId, Address>,
+}
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub enum Status {
+    Spent,
+    Paused,
+}
```
[the remark this answers is not in the document]
Comment (kalabukdima) [reply src/gateway_allocations/rate_limiter.rs:17]: That's what I tried first, but then realized that the presence of `retry_after` is independent from the result status.
In your suggestion I find it strange to encode no value as `0`, so it actually means
```rust
pub enum SpendStatus {
    NoAllocation,
    Paused(Duration),
    Spent(Option<Duration>)
}
```

I thought it's a bit too verbose, but if you're ok with it, I'll gladly make it like this

File: src/query/result.rs [hunk 2]
```diff
@@ -82,3 +82,29 @@ impl IntoResponse for QueryError {
...
 }
+
+impl From<&QueryError> for sqd_messages::query_error::Err {
+    fn from(e: &QueryError) -> Self {
+        use sqd_messages::query_error::Err;
+        match e {
+            QueryError::NotFound => Err::NotFound(e.to_string()),
+            QueryError::BadRequest(e) => Err::BadRequest(e.clone()),
+            QueryError::ServiceOverloaded => Err::ServerOverloaded(()),
+            QueryError::NoAllocation => Err::TooManyRequests(()),
```
[the remark this answers is not in the document]
Comment (kalabukdima) [reply src/query/result.rs:93]: We were planning to not accept any messages from gateways with too little CU allocated, so the `no allocation` will never be sent in practice.

Do you mean that if the client gets "`too many requests`" error, it should count as a spent CU?

File: src/gateway_allocations/rate_limiter.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (eldargab) [reply src/gateway_allocations/rate_limiter.rs:17]: >In your suggestion I find it strange to encode no value as 0

Time till next possible request is `0`, meaning you can make the request right away 🤷‍♂️

File: src/query/result.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (eldargab) [reply src/query/result.rs:93]: >Do you mean that if the client gets "too many requests" error, it should count as a spent CU?

Yes, that's something that we might consider in the future. 

>We were planning to not accept any messages from gateways with too little CU allocated, so the no allocation will never be sent in practice.

That's not user friendly.

**Pull request merged**